joomla/joomla-cms
# [4.0] Set modified by user to created by user for new items

Pull Request for Issue # .

### Summary of Changes

With the changes for datime database columns recently merged, the handling of modified times was made consistent everywhere so that if something has never been modified, the modified time is equal to the created time (with 1 exception: template overrides, where null for modified time is used to indicate that a component for which an override has been made has meanwhile been uninstalled).

This PR here does the same for the modified by user id, so that at the end if something has never been modified, the modified user id is equal to the created user id.

Together with the first change, the handling is then equal to what filesystems do with files and folder.

Finally, the installer is corrected so it changes the user ID's from hard-coded default to random one also for the workflows.

### Testing Instructions

If you have both MySQL and PostgreSQL, please test on both if possible.

#### Test 1: New installation

1. Have a new installation of the 4.0-dev branch or a 4.0-Alpha-12.
2. Login to backend and switch error reporting to "Maximum" or "Development" in Global Config - Server settings.
3. Create each one new item of following types and don't modify it after having saved it for the first time:

- banner
- category
- article
- contact
- field group
- field
- finder filter
- newsfeed
- tag
- user note
- workflow

4. Check modification times and modified by users in the edit views of the particular items or in database using appropriate tools, and check creator user id and modified by user id for the default workflow.

Result: See section "Actual result" for 4.0 below.

5. Apply the patch of this PR. Attention: It happened to me that the patchtester timed out, so not all changes were applied. To be safe, download the zip of this branch and copy the changed files manually to your J4. If you want to save this work: Download a modified nightly build from the following link which includes this PR [https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Full_Package_2019-10-21_pr-26745.zip](https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Full_Package_2019-10-21_pr-26745.zip), clear your Joomla folder so it is empty, delete all database tables, unzip the download into the Joomla folder and then perform a new installation. Then skip the next step 6 and continue with 7.
6. Delete file `configuration.php` in your Joomla root folder, and delete all Joomla database tables in PhpMyAdmin or PhpPgAdmin (depending on your database type).
7. Repeat steps 1 to 4 above.

Result: See section "Expected result" below.

#### Test 2: Update of a 3.9.12 or staging.

1. Have a new installation of the staging branch or 3.9.12.
2. Login to backend and switch error reporting to "Maximum" or "Development" in Global Config - Server settings.
3. Create each one new item of following types and don't modify it after having saved it for the first time:

- banner
- category
- article
- contact
- field group
- field
- finder filter
- newsfeed
- tag
- user note

4. Check modification times and modified by users in the edit views of the particular items or in database using appropriate tools.

Result: See section "Actual result" for 3.x below.

5. Update to a patched nightly build which is equal to current 4.0-dev plus the changes in this PR using the "Upload & Update" tab of the Joomla Update component and the update container previously downloaded from following URL: [https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-10-21_pr-26745.zip](https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-10-21_pr-26745.zip)
6. Check again modification times and modified by users in the edit views of the particular items or in database using appropriate tools.

Result: See section "Expected result" below.

### Expected result

On 4.0 after patch (new installed and updated): For items which never have been modified, the time of last modification is equal to the time of creation, and the user who has modified it as last is equal to the user who created it.

After new installation, the default workflow has the right creator user id and modified by user id of the super user.

### Actual result

On 4.0 before patch: For items which never have been modified, the time of last modification is equal to the time of creation, but the id of the user who has modified it as last is zero, i.e. the modified by user field is empty in the edit view.

On 3.x or staging before patch: For items which never have been modified, the time of last modification is equal to the (pseudo-)nulldatetime of the database type, and the id of the user who has modified it as last is zero, i.e. the modified by user field is empty in the edit view.

### Documentation Changes Required

None.
